Provide rich hover for Zowe Explorer profiles with connection info

SanthoshiBoyina1 opened the pull request:
## Proposed changes
Solves https://github.com/zowe/zowe-explorer-vscode/issues/3580
The Hover for Zowe Explorer profiles contains the following info:
```
Profile Name,
Profile Type,
Auth method : (Basic / Token / Certificate Authentication / Unknown),
User ID (if using basic auth)
Pattern (For a Dataset Node)
Path (For a Uss Node)
Search Criteria (For a Job Node)
Config File: (Global/ Project)
Auto Store value : (true/ false)
Secure Credentials Enabled : (true/ false)
```



<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->

## Release Notes

<!-- Include the Milestone Number and a small description of your change that will be added to the changelog. -->
<!-- If there is a linked issue, it should have the same milestone as this PR. -->

Milestone:

Changelog:

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [ ] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [ ] There is coverage for the code that I have added
- [ ] I have added new test cases and they are passing
- [ ] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did, what alternatives you considered, etc... -->

zFernand0 commented on packages/zowe-explorer/src/configuration/Profiles.ts:
```diff
@@ -96,10 +96,10 @@ export class Profiles extends ProfilesCache {
         ZoweLogger.trace("Profiles.checkCurrentProfile called.");
         let profileStatus: Validation.IValidationProfile = { name: theProfile.name, status: "unverified" };
-        const usingBasicAuth = theProfile.profile.user && theProfile.profile.password;
+        let usingBasicAuth = theProfile.profile.user && theProfile.profile.password;
```
If we address Copilot's comment, this should be fine to move back to `const`. Unless the intention was to reuse the value in the `else if` condition, in which case, the `if` conditions need to be independent of each other (i.e. no `else`)

zFernand0 commented on packages/zowe-explorer/src/utils/AuthUtils.ts:
```diff
@@ -179,6 +179,103 @@ export class AuthUtils {
+                        toolTipList[authMethodIndex] = `Auth method: Token-based Authentication`;
+                    } else if (usingCertAuth) {
+                        toolTipList[authMethodIndex] = `Auth method: Certificate Authentication`;
+                    } else if (!usingTokenAuth && !usingBasicAuth && !usingCertAuth) {
```
Same as previous comment 😋

zFernand0 commented on packages/zowe-explorer/src/trees/job/JobTree.ts:
It seems that we are missing the `user` from the Jobs hover description 😋 
Should we have it there as well? (for consistency with DS and USS)

zFernand0 reviewed (changes requested):
LGTM! 😋 
---

Love the detailed information on-hover 🙏 
Left a few comments and maybe we can implement some of the Copilot suggestions 😋

zFernand0 commented:
One more thing...
How can we make this configurable? 
For instance, there may be times that I want to hide the user ID (for a demo), or I may want to see the host/ip (when debugging)

phaumer commented:
I leave the review to the code owners, but I tried and love the hover!
One thing I would like to see as well is: profile-type.

JillieBeanSim reviewed (dismissed):
This looks great! thanks for working on this @SanthoshiBoyina1

zFernand0 commented on packages/zowe-explorer/src/utils/AuthUtils.ts:
```diff
@@ -179,6 +179,125 @@ export class AuthUtils {
+                        }
+                        break;
+                    }
+                    case sessionNode.description && (sessionNode.description as string).includes("Owner: "): {
```
I see that we are using backticks (``` ` ```) on previous string definitions, which the prettier (or a linter rule) may advice against.

However, Since we are using it in other places, I do have a quick suggestion/trick that could save some type checking here.

```suggestion
                    case `${sessionNode.description}`.includes("Owner: "): {
```

The above will automatically apply/call the `toString()` method on the `description` property and it should always contain the `includes` since it's already a string 😋 

---

This is not a request for changes since the behavior should be the same as it is. 😋

zFernand0 reviewed (commented):
Thanks for adding tests for the AuthUtils changes. 🙏 

I left some comments regarding the switch(true)-case logic below.

They are mostly suggestions for consideration, thus why I'm not requesting changes.

anaxceron commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Implemented the `SharedActions.refreshProvider` function to allow refreshing a single tree provider. [#3524](https://github.com/zowe/zowe-explorer-vscode/pull/3524)
 - Added an new VS Code command palette option `Zowe Explorer: Update Profile Configuration Schema`, allowing users to manually trigger an update to global and project level configuration schema files with registered Zowe Explorer extender types. [#1642](https://github.com/zowe/zowe-explorer-vscode/issues/1642)
 - Added script that will fetch latest localizations at every build now. [#3573](https://github.com/zowe/zowe-explorer-vscode/pull/3573)
+- Implemented rich hover for Zowe Explorer profiles with connection information. [#3580](https://github.com/zowe/zowe-explorer-vscode/pull/3580)
```
```suggestion
- Implemented rich hover for Zowe Explorer profiles to show connection information. [#3580](https://github.com/zowe/zowe-explorer-vscode/pull/3580)
```

anaxceron reviewed (commented):
Left some suggested edits for the changelog

traeok commented on packages/zowe-explorer/src/utils/AuthUtils.ts:
```diff
@@ -179,6 +179,125 @@ export class AuthUtils {
+                    case Boolean(usingTokenAuth): {
+                        toolTipList.push(`Auth Method: Token-based Authentication`);
+                        break;
+                    }
+                    case Boolean(usingBasicAuth): {
+                        toolTipList.push(`Auth Method: Basic Authentication`);
+                        toolTipList.push(`User: ${profile.profile.user as string}`);
+                        break;
+                    }
+                    case Boolean(usingCertAuth): {
+                        toolTipList.push(`Auth Method: Certificate Authentication`);
+                        break;
+                    }
+                    default: {
+                        toolTipList.push(`Auth Method: Unknown`);
+                        break;
+                    }
+                }
+            } else {
+                switch (true) {
+                    case Boolean(usingTokenAuth): {
+                        toolTipList[authMethodIndex] = `Auth Method: Token-based Authentication`;
+                        break;
+                    }
+                    case Boolean(usingBasicAuth): {
+                        toolTipList[authMethodIndex] = `Auth Method: Basic Authentication`;
+                        const userIDIndex = toolTipList.findIndex((key) => key.startsWith("User: "));
+                        if (userIDIndex !== -1) {
+                            toolTipList[userIDIndex] = `User: ${profile.profile.user as string}`;
+                        } else {
+                            toolTipList.splice(authMethodIndex + 1, 0, `User: ${profile.profile.user as string}`);
+                        }
+                        break;
+                    }
+                    case Boolean(usingCertAuth): {
+                        toolTipList[authMethodIndex] = `Auth Method: Certificate Authentication`;
+                        break;
+                    }
+                    default: {
+                        toolTipList[authMethodIndex] = `Auth Method: Unknown`;
```
Could this logic be simplified to avoid duplication? It seems like there is a good amount of overlap with these strings and these strings are hardcoded multiple times.

Also, I think we should localize these strings to support future translation.

traeok reviewed (commented):
Thanks @SanthoshiBoyina1! This looks great so far, I do have a couple requests regarding duplication in tests and localizing strings.

traeok reviewed (changes requested):
See above comments

traeok commented on packages/zowe-explorer/src/configuration/Profiles.ts:
```diff
@@ -152,6 +152,36 @@ export class Profiles extends ProfilesCache {
         }
 
+        if (node !== undefined) {
+            const toolTipList = (node.tooltip as string).split("\n");
```
```suggestion
            const toolTipList = node.tooltip?.split("\n") ?? [];
```

traeok commented on packages/zowe-explorer/src/configuration/Profiles.ts:
```diff
@@ -152,6 +152,36 @@ export class Profiles extends ProfilesCache {
+            const layers = await this.uniqueExistingLayers();
+            const configFileIndex = toolTipList.findIndex((key) => key.startsWith(vscode.l10n.t("Config File: ")));
+            if (configFileIndex === -1) {
+                toolTipList.push(`${vscode.l10n.t("Config File: ")}${layers[0].global ? "Global" : "Project"}`);
+            } else {
+                toolTipList[configFileIndex] = `${vscode.l10n.t("Config File: ")}${layers[0].global ? "Global" : "Project"}`;
```
Sorry to be picky with localizing strings... but the "Global" and "Project" strings should also be wrapped in `vscode.l10n.t`

JillieBeanSim reviewed (approved):
LGTM thanks for addressing feedback @SanthoshiBoyina1

likhithanimma1 reviewed (approved):
Tested works well and LGTM. Thankyou @SanthoshiBoyina1 for the new feature :)

traeok reviewed (changes requested):
Thanks for addressing all the feedback so far Santhoshi! I found a bug where after updating credentials, the profile still shows the old username on the rich hover:

![ZE-RichHover-UpdateCreds](https://github.com/user-attachments/assets/70d8675b-965a-4cba-8989-fa500b3149e5)

SanthoshiBoyina1 commented:
> Thanks for addressing all the feedback so far Santhoshi! I found a bug where after updating credentials, the profile still shows the old username on the rich hover:
> 
> ![ZE-RichHover-UpdateCreds](https://private-user-images.githubusercontent.com/86500639/435717655-70d8675b-965a-4cba-8989-fa500b3149e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDk3MjksIm5iZiI6MTc0NTI0OTQyOSwicGF0aCI6Ii84NjUwMDYzOS80MzU3MTc2NTUtNzBkODY3NWItOTY1YS00Y2JhLTg5ODktZmE1MDBiMzE0OWU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE1MzAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMTBjOWNhYWZjODM3M2Y2ZTljZTUxYjA4OTY4MWU5MzVjNTgyMGQzMzQzOTVkNWVmYjM3ZmZmN2E1NDdiZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uvNjqq1xmKEGVLDXa0GRhFf93tgmDGBhG2sqTslq9go) [ ![ZE-RichHover-UpdateCreds](https://private-user-images.githubusercontent.com/86500639/435717655-70d8675b-965a-4cba-8989-fa500b3149e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDk3MjksIm5iZiI6MTc0NTI0OTQyOSwicGF0aCI6Ii84NjUwMDYzOS80MzU3MTc2NTUtNzBkODY3NWItOTY1YS00Y2JhLTg5ODktZmE1MDBiMzE0OWU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE1MzAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMTBjOWNhYWZjODM3M2Y2ZTljZTUxYjA4OTY4MWU5MzVjNTgyMGQzMzQzOTVkNWVmYjM3ZmZmN2E1NDdiZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uvNjqq1xmKEGVLDXa0GRhFf93tgmDGBhG2sqTslq9go) ](https://private-user-images.githubusercontent.com/86500639/435717655-70d8675b-965a-4cba-8989-fa500b3149e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDk3MjksIm5iZiI6MTc0NTI0OTQyOSwicGF0aCI6Ii84NjUwMDYzOS80MzU3MTc2NTUtNzBkODY3NWItOTY1YS00Y2JhLTg5ODktZmE1MDBiMzE0OWU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE1MzAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMTBjOWNhYWZjODM3M2Y2ZTljZTUxYjA4OTY4MWU5MzVjNTgyMGQzMzQzOTVkNWVmYjM3ZmZmN2E1NDdiZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uvNjqq1xmKEGVLDXa0GRhFf93tgmDGBhG2sqTslq9go) [ ](https://private-user-images.githubusercontent.com/86500639/435717655-70d8675b-965a-4cba-8989-fa500b3149e5.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDk3MjksIm5iZiI6MTc0NTI0OTQyOSwicGF0aCI6Ii84NjUwMDYzOS80MzU3MTc2NTUtNzBkODY3NWItOTY1YS00Y2JhLTg5ODktZmE1MDBiMzE0OWU1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDE1MzAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMTBjOWNhYWZjODM3M2Y2ZTljZTUxYjA4OTY4MWU5MzVjNTgyMGQzMzQzOTVkNWVmYjM3ZmZmN2E1NDdiZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uvNjqq1xmKEGVLDXa0GRhFf93tgmDGBhG2sqTslq9go)

Thanks for testing @traeok , I will look into it now.

zFernand0 reviewed (approved):
LGTM! 😋 
---

I'm really happy with the functionality as-is! 🥳 
I'm happy to re-approve once Trae's comment is addressed. 🙏 

Or if we feel that it might take a bit longer... we can always get this merged in, and address the issue during the following week(s) of testing! 🙏 

Let us know, Santoshi! 🙏

SanthoshiBoyina1 commented:
> ation after updating the credentials.

I am bit confused on deciding which would be the right approach. And I would go with addressing this issue during the following weeks of testing @zFernand0 .

traeok commented:
> I am bit confused on deciding which would be the right approach. And I would go with addressing this issue during the following weeks of testing @zFernand0 .

@SanthoshiBoyina1 Would it be possible to update the tooltip after entering in new credentials? I don't want to hold up this PR if the change is more involved, but it may confuse users who run into issues with authentication. If I set an invalid username and list data sets, the operation fails and I get the Update Credentials prompt, which is expected. However, if I take a look at the tooltip after the request fails, the rich hover still shows the username that previously worked.

If we update the username in the rich hover after the user updates their credentials, that would resolve this desync issue.

traeok reviewed (approved):
Thanks Santhoshi, this looks awesome! I appreciate the discussion regarding the username desync.

SanthoshiBoyina1 commented:
> hink that the trees are out of sync, when the most up-to-date credentials are actually in use.

I will address this username desync issue in a separate PR @traeok @zFernand0
